#### 1.  Iterate through htm files and strip html and save as a text file

#### 2. Clean text

In [6]:
import os
import re

path = 'data/'
arr = os.listdir(path)
for filename in arr:
    if filename.endswith('.txt'):
        xpath = os.path.join(path,filename)
        with open(xpath, 'r') as f:
                text = f.read()
                reclean = re.compile('<.*?>')
                htmlclean = re.sub(reclean,'', text)
                cleantext = " ".join(htmlclean.split())
                outFile = open(path + os.path.splitext(filename)[0] +".txt",'w')
                outFile.write(cleantext)
                outFile.close()   

Single File test

In [7]:
import os
import pandas as pd
import spacy
from spacy.tokenizer import Tokenizer
from spacy.lang.en import English # Import the English language class

nlp = spacy.load('en_core_web_sm')
#nlp = English()
#nlp = spacy.laod('en')

path = 'data/'
arr = os.listdir(path)
for filename in arr:
    if filename.endswith('.txt'):
        xpath = os.path.join(path,filename)
        with open(xpath, 'r') as file:
            contents = file.read()
            doc = nlp(contents)
            entities = [(ent.label_, ent.text, filename) for ent in doc.ents]
            df = pd.DataFrame(entities, columns =['type', 'text', 'filename']) 
            

In [8]:
df

type                                              text  \
0         ORG                                               SEC   
1        DATE                                      July 5, 2017   
2         GPE                                             Japan   
3         ORG  the Japan Bank for International Cooperation Act   
4     PERCENT                                            1.750%   
..        ...                                               ...   
800  CARDINAL                                       6,913 6,910   
801       ORG                                         Education   
802       ORG                                           Science   
803  CARDINAL                                       6,145 5,846   
804  CARDINAL                                       5,598 5,703   

                    filename  
0    424B5_d844216d424b5.txt  
1    424B5_d844216d424b5.txt  
2    424B5_d844216d424b5.txt  
3    424B5_d844216d424b5.txt  
4    424B5_d844216d424b5.txt  
..                       ...  
800  424B5_d844216d424b5.txt  
801  424B5_d844216d424b5.txt  
802  424B5_d844216d424b5.txt  
803  424B5_d844216d424b5.txt  
804  424B5_d844216d424b5.txt  

[805 rows x 3 columns]

In [134]:
import os
import pandas as pd
import spacy
from spacy.tokenizer import Tokenizer
from spacy.lang.en import English # Import the English language class
from spacy.matcher import Matcher

nlp = spacy.load('en_core_web_sm')
#nlp = English()
#nlp = spacy.laod('en')

path = 'data/'
arr = os.listdir(path)
for filename in arr:
    if filename.endswith('.txt'):
        xpath = os.path.join(path,filename)
        with open(xpath, 'r') as file:
            contents = file.read()
            matcher = Matcher(nlp.vocab)
            pattern = [{'LOWER': 'cusip'},{}]
            matcher.add("CUSIP", None, pattern)
            doc = nlp(contents)
            matches = matcher(doc)
            print(matches)

[]
[]
[(15374025435414067454, 125, 127)]
[(15374025435414067454, 127, 129)]
[(15374025435414067454, 127, 129)]
[(15374025435414067454, 228, 230)]
[(15374025435414067454, 137, 139)]
[(15374025435414067454, 137, 139)]
[(15374025435414067454, 247, 249), (15374025435414067454, 722, 724), (15374025435414067454, 1440, 1442)]
[(15374025435414067454, 2110, 2112), (15374025435414067454, 2361, 2363)]
[(15374025435414067454, 185, 187)]
[(15374025435414067454, 6305, 6307)]
[(15374025435414067454, 2110, 2112), (15374025435414067454, 2361, 2363)]
[(15374025435414067454, 5194, 5196), (15374025435414067454, 7826, 7828)]
[(15374025435414067454, 111, 113)]
[(15374025435414067454, 772, 774), (15374025435414067454, 14148, 14150), (15374025435414067454, 14501, 14503), (15374025435414067454, 14537, 14539)]
[(15374025435414067454, 771, 773)]
[(15374025435414067454, 768, 770), (15374025435414067454, 14028, 14030), (15374025435414067454, 14377, 14379), (15374025435414067454, 14413, 14415)]
[(153740254354140674

#### Using Regex to extract CUSIP and Issuer to an Excel file

In [135]:
import re
import os
import pandas as pd

path = 'data/'
files=[i for i in os.listdir(path) if i.endswith("txt")]

w={'FILE':[],'CUSIP': [], 'ISSUER': []}

for file in files:

    f=open(path+ file)
    data=f.read()
    a=re.findall(r'[\n]?[_-]?\s+(?:[_-]{9,})?[\s\r\t\n]*\(CUSIP Number\)',data)
    a1="".join(a).replace('\n',' ')
    b=re.findall(r'Issuer(.*?)issuer',data,re.DOTALL)
    b1="".join(b).replace('\n',' ')
     #b=re.findall(r'CUSIP(.*?)cusip',data,re.DOTALL)

    loc= os.path.join(file)
    
    w['FILE'].append(loc)
    w['CUSIP'].append(a1)
    w['ISSUER'].append(b1) 
    #print (os.path.join(file))

df=pd.DataFrame(data=w)
df.to_excel('data/return.xlsx')

#### Using NLTK to assess poatterns of words associated with tagged list for Callable

APPROACH:  The problem could be viewed as a supervised learning problem where I assist by tagging the callables and then train them recognize callable using scikit learn

In [1]:
import nltk
nltk.download('popular')

[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to
[nltk_data]    |     C:\Users\Lucky\AppData\Roaming\nltk_data...
[nltk_data]    |   Package cmudict is already up-to-date!
[nltk_data]    | Downloading package gazetteers to
[nltk_data]    |     C:\Users\Lucky\AppData\Roaming\nltk_data...
[nltk_data]    |   Package gazetteers is already up-to-date!
[nltk_data]    | Downloading package genesis to
[nltk_data]    |     C:\Users\Lucky\AppData\Roaming\nltk_data...
[nltk_data]    |   Package genesis is already up-to-date!
[nltk_data]    | Downloading package gutenberg to
[nltk_data]    |     C:\Users\Lucky\AppData\Roaming\nltk_data...
[nltk_data]    |   Package gutenberg is already up-to-date!
[nltk_data]    | Downloading package inaugural to
[nltk_data]    |     C:\Users\Lucky\AppData\Roaming\nltk_data...
[nltk_data]    |   Package inaugural is already up-to-date!
[nltk_data]    | Downloading package movie_reviews to
[nltk_data]   

True

In [5]:
import pandas as pd
import os
import nltk
from nltk.tokenize import PunktSentenceTokenizer
from nltk.corpus import reuters

tokenizer = PunktSentenceTokenizer()
path = 'data/'
arr = os.listdir(path)
data = []
for filename in arr:
    if filename.endswith('.txt'):
        xpath = os.path.join(path,filename)
        with open(xpath, 'r') as file:
            contents = file.read()
            data.extend(tokenizer.tokenize(contents))
            

In [6]:
df = pd.DataFrame(data, columns =['text']) 
df.to_excel('data/sentences.xlsx')

In [95]:
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
from sklearn.base import TransformerMixin
from sklearn.pipeline import Pipeline

In [8]:
data = pd.read_csv('data\callable_text.csv')
data.head()

text  label
0  Such factors may have the effect of discouragi...      0
1  If 3-month U.S. dollar LIBOR is discontinued o...      0
2  The notes differ from conventional fixed-rate ...      0
3  Validity of the Notes   In the opinion of Davi...      0
4  “NY Federal Reserve’s website” means the websi...      0

In [9]:
data.shape #Only 849 total sentences evaluated 

(849, 2)

In [10]:
data.info()  # marker was manually marked

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 849 entries, 0 to 848
Data columns (total 2 columns):
text     849 non-null object
label    849 non-null int64
dtypes: int64(1), object(1)
memory usage: 13.3+ KB


In [11]:
data.label.value_counts() # 1 indicates text associated with callable

0    754
1     95
Name: label, dtype: int64

##### Tokening the data using spacy

In [12]:
import spacy
import string

from spacy.lang.en.stop_words import STOP_WORDS
from spacy.tokenizer import Tokenizer
from spacy.lang.en import English # Import the English language class
from spacy.matcher import Matcher

In [13]:
punctuations = string.punctuation# Create our list of punctuation marks
nlp = spacy.load('en') # Create list of stopwords
stop_words = spacy.lang.en.stop_words.STOP_WORDS
parser = English()# Load English tokenizer, tagger, parser, NER and word vectors


def spacy_tokenizer(sentence):# Creating tokenizer function
    
    mytokens = parser(sentence)# Creating token object, used to create documents with linguistic annotations.
    mytokens = [ word.lemma_.lower().strip() if word.lemma_ != "-PRON-" else word.lower_ for word in mytokens ] # Lemmatizing each token and converting each token into lowercase
    mytokens = [ word for word in mytokens if word not in stop_words and word not in punctuations ] # Removing stop words
    
    return mytokens# return preprocessed list of tokens


In [14]:
# Custom transformer using spaCy
class predictors(TransformerMixin):
    def transform(self, X, **transform_params):
        # Cleaning Text
        return [clean_text(text) for text in X]

    def fit(self, X, y=None, **fit_params):
        return self

    def get_params(self, deep=True):
        return {}

# Basic function to clean the text
def clean_text(text):
    # Removing spaces and converting text into lowercase
    return text.strip().lower()

In [22]:
bow_vector = CountVectorizer(tokenizer = spacy_tokenizer, ngram_range=(1,1))

In [20]:
tfidf_vector = TfidfVectorizer(tokenizer = spacy_tokenizer)

In [24]:
from sklearn.model_selection import train_test_split

X = data['text'] # the features we want to analyze
ylabels = data['label'] # the labels, or answers, we want to test against

X_train, X_test, y_train, y_test = train_test_split(X, ylabels, test_size=0.3)

### Logistic Regression Classifier

In [25]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()

# Create pipeline using Bag of Words
pipe = Pipeline([('cleaner', predictors()),
                 ('vectorizer', bow_vector),
                 ('classifier', classifier)])

# model generation
pipe.fit(X_train,y_train)

C:\Users\Lucky\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Pipeline(memory=None,
         steps=[('cleaner', <__main__.predictors object at 0x000001D3CD3021D0>),
                ('vectorizer',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=<function spacy_tokenizer at 0x000001D3C86E2950>,
                                 vocabulary=None)),
                ('classifier',
                 LogisticRegression(C=1.0, class_weight=None, dual=False,
                                    fit_intercept=True, intercept_

In [26]:
from sklearn import metrics
# Predicting with a test dataset
predicted = pipe.predict(X_test)

# Model Accuracy
print("Logistic Regression Accuracy:",metrics.accuracy_score(y_test, predicted))
print("Logistic Regression Precision:",metrics.precision_score(y_test, predicted))
print("Logistic Regression Recall:",metrics.recall_score(y_test, predicted))

Logistic Regression Accuracy: 0.9490196078431372
Logistic Regression Precision: 0.9545454545454546
Logistic Regression Recall: 0.6363636363636364


In other words, overall, our model correctly identified sentence sentiment 94.9% of the time. When it predicted a "callible", that classification was actually callable 96% of the time. When handed an unseen callable sentence, our model identified it as positive 68.4% of the time (just ok -- better than a coin flip)

### Classify Documents

In [60]:
#### Build a dataframe from the files in the data folder and add a callable column

In [61]:
import os
import pandas as pd

data=[]
path = "data/Train/"

for filename in sorted(os.listdir('data/Train')):
    if filename.endswith('.txt'):
        xpath = os.path.join(path,filename)
        with open(xpath, 'r') as file:
            contents = file.read()
            data.append((filename,contents))
filedf = pd.DataFrame(data, columns=['filename','contents'])
filedf['callable'] = ''
filedf = filedf[['callable', 'filename', 'contents']]
filedf.to_csv('data/train.csv')
filedf

callable                              filename  \
0                       424B2_d833231d424b2_C.txt   
1                       424B2_d859033d424b2_N.txt   
2              424B2_dp114411_424b2-mtn1197_N.txt   
3                  424B2_dp115966_424b2-255_N.txt   
4                  424B2_dp116691_424b2-257_N.txt   
5                  424B2_dp120183_424b2-302_N.txt   
6                        424B2_g110203424b2_N.txt   
7                            424B2_gs-424b2_N.txt   
8            424B3_pricingsupplement2-yrfrn_N.txt   
9                       424B5_d832846d424b5_C.txt   
10                      424B5_d834717d424b5_C.txt   
11                      424B5_d844216d424b5_C.txt   

                                             contents  
0   We may, at our option, redeem the floating rat...  
1   We may redeem any series of the notes at our o...  
2   CALCULATION OF REGISTRATION FEE Title of each ...  
3   Citigroup Global Markets Holdings Inc. Novembe...  
4   Citigroup Global Markets Holdings Inc. Novembe...  
5   Citigroup Global Markets Holdings Inc. January...  
6   Royal Bank of Canada Filed Pursuant to Rule 42...  
7   gs-424b2.htm Filed Pursuant to Rule 424(b)(2) ...  
8   Document CALCULATION OF REGISTRATION FEE Title...  
9   If at any time the Bank shall become obligated...  
10  We may redeem the notes at our option, in whol...  
11  Unless we redeem the notes earlier, the notes ...

In [62]:
import pandas as pd
import os
import nltk
from nltk.tokenize import PunktSentenceTokenizer

tokenizer = PunktSentenceTokenizer()
path = 'data/Train'
arr = os.listdir(path)
data = []
filenames = []
for filename in arr:
    if filename.endswith('.txt'):
        xpath = os.path.join(path,filename)
        with open(xpath, 'r') as file:
            contents = file.read()
            data.append(tokenizer.tokenize(contents))
            filenames.append(filename)
training = pd.DataFrame({'filename':filenames, 'data': data})

In [63]:
training

filename  \
0              424B2_d833231d424b2_C.txt   
1              424B2_d859033d424b2_N.txt   
2     424B2_dp114411_424b2-mtn1197_N.txt   
3         424B2_dp115966_424b2-255_N.txt   
4         424B2_dp116691_424b2-257_N.txt   
5         424B2_dp120183_424b2-302_N.txt   
6               424B2_g110203424b2_N.txt   
7                   424B2_gs-424b2_N.txt   
8   424B3_pricingsupplement2-yrfrn_N.txt   
9              424B5_d832846d424b5_C.txt   
10             424B5_d834717d424b5_C.txt   
11             424B5_d844216d424b5_C.txt   

                                                 data  
0   [We may, at our option, redeem the floating ra...  
1   [We may redeem any series of the notes at our ...  
2   [CALCULATION OF REGISTRATION FEE Title of each...  
3   [Citigroup Global Markets Holdings Inc., Novem...  
4   [Citigroup Global Markets Holdings Inc., Novem...  
5   [Citigroup Global Markets Holdings Inc., Janua...  
6   [Royal Bank of Canada Filed Pursuant to Rule 4...  
7   [gs-424b2.htm Filed Pursuant to Rule 424(b)(2)...  
8   [Document CALCULATION OF REGISTRATION FEE Titl...  
9   [If at any time the Bank shall become obligate...  
10  [We may redeem the notes at our option, in who...  
11  [Unless we redeem the notes earlier, the notes...

In [64]:
training = training.explode('data')

In [65]:
training.to_csv('data/training.csv', index = False)

In [102]:
training = pd.read_csv('data/Training.csv')
training['callable']= training['filename'].str.contains('C.txt') 
training

filename  \
0     424B2_d833231d424b2_C.txt   
1     424B2_d833231d424b2_C.txt   
2     424B2_d833231d424b2_C.txt   
3     424B2_d833231d424b2_C.txt   
4     424B2_d833231d424b2_C.txt   
...                         ...   
2739  424B5_d844216d424b5_C.txt   
2740  424B5_d844216d424b5_C.txt   
2741  424B5_d844216d424b5_C.txt   
2742  424B5_d844216d424b5_C.txt   
2743  424B5_d844216d424b5_C.txt   

                                                   data  callable  
0     We may, at our option, redeem the floating rat...      True  
1     For the avoidance of doubt, if the floating ra...      True  
2     Further, if fewer than all of the floating rat...      True  
3     In addition to the optional make-whole redempt...      True  
4     For the avoidance of doubt, if the fixed/float...      True  
...                                                 ...       ...  
2739  The relevant prospectus supplement will specif...      True  
2740  Any notice of redemption of debt securities of...      True  
2741  In the case of a partial redemption, the trust...      True  
2742  We or any of our subsidiaries may at any time ...      True  
2743  Any debt securities of any such series purchas...      True  

[2744 rows x 3 columns]

In [67]:
tokenizer = PunktSentenceTokenizer()
path = 'data/Test'
arr = os.listdir(path)
data = []
filenames = []
for filename in arr:
    if filename.endswith('.txt'):
        xpath = os.path.join(path,filename)
        with open(xpath, 'r') as file:
            contents = file.read()
            data.append(tokenizer.tokenize(contents))
            filenames.append(filename)
test = pd.DataFrame({'filename':filenames, 'data': data})

In [68]:
test

filename  \
0     424B2_a19-18449_4424b2_N.txt   
1       424B2_a19-18449_5424b2.txt   
2       424B2_a19-18449_6424b2.txt   
3       424B2_a19-18959_1424b2.txt   
4        424B2_a20-1355_3424b2.txt   
5        424B2_a20-1355_4424b2.txt   
6            424B2_d26200424b2.txt   
7        424B2_d795231d424b2_C.txt   
8        424B2_d816260d424b2_N.txt   
9        424B2_d881651d424b2_C.txt   
10  424B2_dp120926_424b2-k1431.txt   
11  424B2_dp121287_424b2-mtnps.txt   
12        424B2_efc20-74_424b2.txt   
13       424B2_j1022190424b2_N.txt   
14      424B5_a2240370z424b5_C.txt   

                                                 data  
0   [PROSPECTUS and PRICING SUPPLEMENT NO., 34 PRO...  
1   [PROSPECTUS and PRICING SUPPLEMENT NO., 35 PRO...  
2   [PROSPECTUS and PRICING SUPPLEMENT NO., 36 PRO...  
3   [PROSPECTUS Dated March 14, 2018 and PRICING S...  
4   [PROSPECTUS and PROSPECTUS SUPPLEMENT, each Da...  
5   [PROSPECTUS and PROSPECTUS SUPPLEMENT, each Da...  
6   [Royal Bank of Canada Filed Pursuant to Rule 4...  
7   [424B2 Table of Contents This filing is made p...  
8   [Pricing Supplement Table of Contents This fil...  
9   [424B2 Table of Contents CALCULATION OF REGIST...  
10  [Pricing Supplement No., K1431 To the Underlyi...  
11  [CALCULATION OF REGISTRATION FEE Title of each...  
12  [PROfilePageNumberReset%Num%1%PS-%% CALCULATIO...  
13  [Royal Bank of Canada Filed Pursuant to Rule 4...  
14  [Use these links to rapidly review the documen...

In [186]:
test = test.explode('data')

In [70]:
test.to_csv('data/Test.csv', index = False)

In [147]:
test = pd.read_csv('data/Test.csv')
test['probability'] =""
test['callable']= ""
test

filename  \
0     424B2_a19-18449_4424b2_N.txt   
1     424B2_a19-18449_4424b2_N.txt   
2     424B2_a19-18449_4424b2_N.txt   
3     424B2_a19-18449_4424b2_N.txt   
4     424B2_a19-18449_4424b2_N.txt   
...                            ...   
4554    424B5_a2240370z424b5_C.txt   
4555    424B5_a2240370z424b5_C.txt   
4556    424B5_a2240370z424b5_C.txt   
4557    424B5_a2240370z424b5_C.txt   
4558    424B5_a2240370z424b5_C.txt   

                                                   data probability callable  
0                 PROSPECTUS and PRICING SUPPLEMENT NO.                       
1     34 PROSPECTUS SUPPLEMENT, each Dated September...                       
2                          333-217193 by Supplement No.                       
3     1 dated June 27, 2018 and Filed Pursuant to Ru...                       
4                          2 dated January 4, 2019 U.S.                       
...                                                 ...         ...      ...  
4554  LEGAL OPINIONS Cleary Gottlieb Steen & Hamilto...                       
4555  counsel, and Slaughter and May, our English so...                       
4556  EXPERTS The financial statements incorporated ...                       
4557  40 ZEQ.=4,SEQ=67,EFW="2240411",CP="SANTANDER U...                       
4558  ;5',USER='CHE108057',CD='30-DEC-2019;14:23' TH...                       

[4559 rows x 4 columns]

In [148]:
import spacy
import string

from spacy.lang.en.stop_words import STOP_WORDS
from spacy.tokenizer import Tokenizer
from spacy.lang.en import English # Import the English language class
from spacy.matcher import Matcher

In [149]:
punctuations = string.punctuation# Create our list of punctuation marks
nlp = spacy.load('en') # Create list of stopwords
stop_words = spacy.lang.en.stop_words.STOP_WORDS
parser = English()# Load English tokenizer, tagger, parser, NER and word vectors


def spacy_tokenizer(sentence):# Creating tokenizer function
    
    mytokens = parser(sentence)# Creating token object, used to create documents with linguistic annotations.
    mytokens = [ word.lemma_.lower().strip() if word.lemma_ != "-PRON-" else word.lower_ for word in mytokens ] # Lemmatizing each token and converting each token into lowercase
    mytokens = [ word for word in mytokens if word not in stop_words and word not in punctuations ] # Removing stop words
    
    return mytokens# return preprocessed list of tokens


#### Vectorize DataFrames (Not used)

In [ ]:
##### Not used
def computeTFDIF(tfBoow,idfs):
    tfidf = {}
    for sentence, val in tfBow.items():
        tfdif(sentence) = val*idfs(sentence
    return tfdif

In [88]:
from sklearn.base import TransformerMixin
from sklearn.pipeline import make_pipeline
from sklearn.feature_extraction import DictVectorizer

class RowIterator(TransformerMixin):
    """ Prepare dataframe for DictVectorizer """
    def fit(self, X, y=None):
        return self

    def transform(self, X):
        return (row for _, row in X.iterrows())

vectorizer = make_pipeline(RowIterator(), DictVectorizer())

In [89]:
# now we can use vectorizer 
vectorizer.fit_transform(training[['data']])

<2744x2057 sparse matrix of type '<class 'numpy.float64'>'
	with 2744 stored elements in Compressed Sparse Row format>

In [90]:
vectorizer.transform(test[['data']])

<4559x2057 sparse matrix of type '<class 'numpy.float64'>'
	with 412 stored elements in Compressed Sparse Row format>

### Vectorize  and Train

In [162]:
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
from sklearn.base import TransformerMixin
from sklearn.pipeline import Pipeline

In [163]:
training = pd.read_csv('data/Training.csv')
training['callable']= training['filename'].str.contains('C.txt') 
training

filename  \
0     424B2_d833231d424b2_C.txt   
1     424B2_d833231d424b2_C.txt   
2     424B2_d833231d424b2_C.txt   
3     424B2_d833231d424b2_C.txt   
4     424B2_d833231d424b2_C.txt   
...                         ...   
2739  424B5_d844216d424b5_C.txt   
2740  424B5_d844216d424b5_C.txt   
2741  424B5_d844216d424b5_C.txt   
2742  424B5_d844216d424b5_C.txt   
2743  424B5_d844216d424b5_C.txt   

                                                   data  callable  
0     We may, at our option, redeem the floating rat...      True  
1     For the avoidance of doubt, if the floating ra...      True  
2     Further, if fewer than all of the floating rat...      True  
3     In addition to the optional make-whole redempt...      True  
4     For the avoidance of doubt, if the fixed/float...      True  
...                                                 ...       ...  
2739  The relevant prospectus supplement will specif...      True  
2740  Any notice of redemption of debt securities of...      True  
2741  In the case of a partial redemption, the trust...      True  
2742  We or any of our subsidiaries may at any time ...      True  
2743  Any debt securities of any such series purchas...      True  

[2744 rows x 3 columns]

In [164]:
training.callable.value_counts()

False    2693
True       51
Name: callable, dtype: int64

In [165]:
import spacy
import string

from spacy.lang.en.stop_words import STOP_WORDS
from spacy.tokenizer import Tokenizer
from spacy.lang.en import English # Import the English language class
from spacy.matcher import Matcher

In [166]:
# Custom transformer using spaCy
class predictors(TransformerMixin):
    def transform(self, X, **transform_params):
        # Cleaning Text
        return [clean_text(text) for text in X]

    def fit(self, X, y=None, **fit_params):
        return self

    def get_params(self, deep=True):
        return {}

# Basic function to clean the text
def clean_text(text):
    # Removing spaces and converting text into lowercase
    return text.strip().lower()

In [167]:
bow_vector = CountVectorizer(tokenizer = spacy_tokenizer, ngram_range=(1,1))

In [168]:
tfidf_vector = TfidfVectorizer(tokenizer = spacy_tokenizer)

In [169]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()

# Create pipeline using Bag of Words
pipe = Pipeline([('cleaner', predictors()),
                 ('vectorizer', bow_vector),
                 ('classifier', classifier)])

# model generation
pipe.fit(training['data'],training['callable'])

Pipeline(memory=None,
         steps=[('cleaner', <__main__.predictors object at 0x000002BEC3934DA0>),
                ('vectorizer',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=<function spacy_tokenizer at 0x000002BEC288EF28>,
                                 vocabulary=None)),
                ('classifier',
                 LogisticRegression(C=1.0, class_weight=None, dual=False,
                                    fit_intercept=True, intercept_

In [170]:
# model generation
pipe.predict(training['data'])

array([ True,  True,  True, ...,  True,  True,  True])

In [180]:
from sklearn import metrics
from sklearn.metrics import accuracy_score
# Predicting with a test dataset

test['callable'] = pipe.predict_proba(test['data'])[:,1]
Resolve = test.groupby('filename')['callable'].max() 


# Model Accuracy
#print("Logistic Regression Accuracy:",metrics.accuracy_score(y_test, predicted))
#print("Logistic Regression Precision:",metrics.precision_score(y_test, predicted))
#print("Logistic Regression Recall:",metrics.recall_score(y_test, predicted))

In [181]:
Resolve

filename
424B2_a19-18449_4424b2_N.txt      0.068964
424B2_a19-18449_5424b2.txt        0.207525
424B2_a19-18449_6424b2.txt        0.069939
424B2_a19-18959_1424b2.txt        0.121834
424B2_a20-1355_3424b2.txt         0.389613
424B2_a20-1355_4424b2.txt         0.387714
424B2_d26200424b2.txt             0.092872
424B2_d795231d424b2_C.txt         1.000000
424B2_d816260d424b2_N.txt         1.000000
424B2_d881651d424b2_C.txt         1.000000
424B2_dp120926_424b2-k1431.txt    0.161670
424B2_dp121287_424b2-mtnps.txt    0.968512
424B2_efc20-74_424b2.txt          1.000000
424B2_j1022190424b2_N.txt         0.092872
424B5_a2240370z424b5_C.txt        0.999999
Name: callable, dtype: float64

In [184]:
call = pd.DataFrame(Resolve)
call['probability'] = call['callable'].apply(lambda x: 'True' if x >= .75 else 'False')
call

callable probability
filename                                            
424B2_a19-18449_4424b2_N.txt    0.068964       False
424B2_a19-18449_5424b2.txt      0.207525       False
424B2_a19-18449_6424b2.txt      0.069939       False
424B2_a19-18959_1424b2.txt      0.121834       False
424B2_a20-1355_3424b2.txt       0.389613       False
424B2_a20-1355_4424b2.txt       0.387714       False
424B2_d26200424b2.txt           0.092872       False
424B2_d795231d424b2_C.txt       1.000000        True
424B2_d816260d424b2_N.txt       1.000000        True
424B2_d881651d424b2_C.txt       1.000000        True
424B2_dp120926_424b2-k1431.txt  0.161670       False
424B2_dp121287_424b2-mtnps.txt  0.968512        True
424B2_efc20-74_424b2.txt        1.000000        True
424B2_j1022190424b2_N.txt       0.092872       False
424B5_a2240370z424b5_C.txt      0.999999        True